From 208a31200963bbef6627cc09fd7ebf710407e1cd Mon Sep 17 00:00:00 2001
From: Keith Busch <kbusch@kernel.org>
Date: Fri, 10 Dec 2021 18:22:53 +0800
Subject: [PATCH] block: return errors from blk_execute_rq()
Git-commit: fb9b16e15cd70e21d8af7f03d700deb9509c2ce8
Patch-mainline: v5.14-rc1
References: bsn#12
openEuler-commit: 208a31200963bbef6627cc09fd7ebf710407e1cd


The synchronous blk_execute_rq() had not provided a way for its callers
to know if its request was successful or not. Return the blk_status_t
result of the request.

Reviewed-by: Christoph Hellwig <hch@lst.de>
Reviewed-by: Ming Lei <ming.lei@redhat.com>
Signed-off-by: Keith Busch <kbusch@kernel.org>
Reviewed-by: Chaitanya Kulkarni <chaitanya.kulkarni@wdc.com>
Link: https://lore.kernel.org/r/20210610214437.641245-4-kbusch@kernel.org
Signed-off-by: Jens Axboe <axboe@kernel.dk>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 block/blk-exec.c       | 7 +++++--
 include/linux/blkdev.h | 4 ++--
 2 files changed, 7 insertions(+), 4 deletions(-)

--- a/block/blk-exec.c
+++ b/block/blk-exec.c
@@ -21,7 +21,7 @@ static void blk_end_sync_rq(struct reque
 {
 	struct completion *waiting = rq->end_io_data;
 
-	rq->end_io_data = NULL;
+	rq->end_io_data = (void *)(uintptr_t)error;
 
 	/*
 	 * complete last, if this is a stack request the process (and thus
@@ -73,8 +73,9 @@ EXPORT_SYMBOL_GPL(blk_execute_rq_nowait)
  * Description:
  *    Insert a fully prepared request at the back of the I/O scheduler queue
  *    for execution and wait for completion.
+ *    Return: The blk_status_t result provided to blk_mq_end_request().
  */
-void blk_execute_rq(struct request_queue *q, struct gendisk *bd_disk,
+blk_status_t blk_execute_rq(struct request_queue *q, struct gendisk *bd_disk,
 		   struct request *rq, int at_head)
 {
 	DECLARE_COMPLETION_ONSTACK(wait);
@@ -89,5 +90,7 @@ void blk_execute_rq(struct request_queue
 		while (!wait_for_completion_io_timeout(&wait, hang_check * (HZ/2)));
 	else
 		wait_for_completion_io(&wait);
+
+	return (blk_status_t)(uintptr_t)rq->end_io_data;
 }
 EXPORT_SYMBOL(blk_execute_rq);
--- a/include/linux/blkdev.h
+++ b/include/linux/blkdev.h
@@ -956,10 +956,10 @@ extern int blk_rq_map_kern(struct reques
 extern int blk_rq_map_user_iov(struct request_queue *, struct request *,
 			       struct rq_map_data *, const struct iov_iter *,
 			       gfp_t);
-extern void blk_execute_rq(struct request_queue *, struct gendisk *,
-			  struct request *, int);
 extern void blk_execute_rq_nowait(struct gendisk *,
 				  struct request *, int, rq_end_io_fn *);
+blk_status_t blk_execute_rq(struct request_queue *, struct gendisk *,
+			    struct request *, int);
 
 /* Helper to convert REQ_OP_XXX to its string format XXX */
 extern const char *blk_op_str(unsigned int op);
